main/lsp/requests/rename.cc

```diff
@@ -212,22 +228,32 @@ variant<JSONNullObject, unique_ptr<WorkspaceEdit>> RenameTask::getRenameEdits(LS
             }
         }
     } else {
-        renamer = make_unique<ConstRenamer>(originalName, newName);
+        renamer = make_unique<ConstRenamer>(gs, config, originalName, newName);
         symbolsToRename.push_back(symbol);
     }
 
     for (auto sym : symbolsToRename) {
-        vector<unique_ptr<Location>> references = getReferencesToSymbol(typechecker, sym);
+        // vector<unique_ptr<Location>> references = getReferencesToSymbol(typechecker, sym);
+
+        auto queryResult = queryBySymbol(typechecker, sym);
+        if (queryResult.error) {
+            return JSONNullObject();
+        }
 
-        for (auto &location : references) {
-            // Get text at location.
-            auto fref = config.uri2FileRef(gs, location->uri);
-            if (fref.data(gs).isPayload()) {
+        // We want location but also the type of the expression at that location; and for some expression types like
+        // sends, we need more than just a location, for parsing purposes (the send location is too broad and makes us
+        // parse too much).
+        // TODO(soam): do we need to dedup locations?
+        for (auto &response : queryResult.responses) {
```

Comment:
Note that `getReferencesToSymbol` prunes Send results from untyped files and dedupes locations. I suspect that you will want to dedupe locations. DSL passes like `attr_writer` are likely to trigger the dedupe logic as they insert synthetic AST nodes that blame to the same source line.